File: src/types/automation.ts
export type NodeStatus = 'idle' | 'processing' | 'completed' | 'failed' | 'blocked';

export interface AutomationTelemetryEvent {
  automation_id: string;
  node_id: string;
  status: NodeStatus;
  output?: unknown;
  error?: string;
}

export interface ScheduleDefinition {
  schedule_type: "interval" | "daily" | "weekly" | "monthly" | "specific_dates" | "one_time";
  interval_minutes?: number;
  time_of_day?: string;           // "HH:MM"
  days_of_week?: string[];        // ["Mon","Tue","Fri"]
  repeat_every?: number;          // every N weeks (default 1)
  days_of_month?: number[];       // [1, 15] for monthly
  specific_dates?: string[];      // ["2026-05-01"] for specific_dates
  run_at?: string;                // ISO datetime for one_time
  end_condition?: "never" | "on_date" | "after_occurrences";
  end_date?: string;              // YYYY-MM-DD
  max_occurrences?: number;
  occurrence_count?: number;
  active: boolean;
}

export type AutomationBusyPolicy = 'wait' | 'queue' | 'skip' | 'fail';
export type AutomationAgentConversation = 'current' | 'fresh_background';

export type AutomationRoleAssignment =
  | {
      target_type: 'agent';
      agent_id: string;
      conversation: AutomationAgentConversation;
      busy_policy?: AutomationBusyPolicy;
    }
  | {
      target_type: 'temporary_provider';
      provider: string;
      workspace?: string;
      model?: string;
      effort?: string;
    };

export type AutomationAssignments = Record<string, AutomationRoleAssignment>;

/** Mirrors the 6b Rust `AutomationSchedule` DTO (a persisted automation invoker). */
export interface AutomationSchedule {
  id: string;
  blueprint_id: string;
  name: string;
  provider?: string | null;
  workspace?: string | null;
  input: unknown;
  bindings: Record<string, string>;
  assignments?: AutomationAssignments;
  schedule: ScheduleDefinition;
  next_run_epoch_ms?: number | null;
  paused_remaining_ms?: number | null;
  is_paused: boolean;
  last_run_status?: string | null;
  last_run_error?: string | null;
  last_run_epoch_ms?: number | null;
}

// ---------------------------------------------------------------------------
// Listener invokers (file watch, inbound webhook, outbound web poll)
// ---------------------------------------------------------------------------

/** What a listener does when it fires while one of its own runs is active. */
export type ListenerOverlapPolicy = 'skip' | 'coalesce' | 'parallel';

export type FileChangeKind = 'created' | 'modified' | 'removed';
export type ListenerWebhookAuth = 'token' | 'hmac_sha256';
export type ListenerPollMethod = 'get' | 'head';
export type ListenerFingerprintSource =
  | 'etag_or_last_modified'
  | 'body_hash'
  | 'json_pointer'
  | 'regex';

export interface FileWatchTrigger {
  type: 'file_watch';
  path: string;
  recursive: boolean;
  patterns: string[];
  ignore: string[];
  events: FileChangeKind[];
  debounce_ms: number;
}

export interface WebhookTrigger {
  type: 'webhook';
  path_segment: string;
  auth: ListenerWebhookAuth;
  signature_header?: string | null;
  max_body_bytes: number;
}

export interface WebPollTrigger {
  type: 'web_poll';
  url: string;
  interval_seconds: number;
  method: ListenerPollMethod;
  headers: Record<string, string>;
  fingerprint: ListenerFingerprintSource;
  json_pointer?: string | null;
  regex?: string | null;
  max_body_bytes: number;
}

export type ListenerTrigger = FileWatchTrigger | WebhookTrigger | WebPollTrigger;

export interface ListenerRejection {
  reason: string;
  at_epoch_ms: number;
}

/**
 * Every app-written field. Kept separate from the trigger config so no field
 * has two writers; the UI reads this and never sends it back.
 */
export interface ListenerRuntime {
  armed: boolean;
  arm_error?: string | null;
  last_fire_epoch_ms?: number | null;
  last_run_status?: string | null;
  last_run_error?: string | null;
  last_rejection?: ListenerRejection | null;
  fire_count: number;
  recent_fire_epoch_ms: number[];
  /** Set when the rate ceiling trips. `enabled` stays the user's own switch. */
  disabled_reason?: string | null;
  poll_fingerprint?: string | null;
  next_poll_epoch_ms?: number | null;
  consecutive_failures: number;
}

/** Mirrors the Rust `AutomationListener` DTO. */
export interface AutomationListener {
  id: string;
  blueprint_id: string;
  name: string;
  enabled: boolean;
  trigger: ListenerTrigger;
  provider?: string | null;
  workspace?: string | null;
  input: unknown;
  bindings: Record<string, string>;
  assignments?: AutomationAssignments;
  overlap?: ListenerOverlapPolicy | null;
  runtime: ListenerRuntime;
}

/** A listener plus derived facts the backend computes but never stores. */
export interface ListenerView extends AutomationListener {
  webhook_url?: string | null;
  /** Whether a secret exists. Never the secret itself. */
  has_secret: boolean;
}

export interface ListenerGatewayConfig {
  schema: number;
  host: string;
  port: number;
}

File: crates/wardian-core/src/models/automation.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(tag = "target_type", rename_all = "snake_case")]
pub enum AutomationRoleAssignment {
    Agent {
        agent_id: String,
        #[serde(default = "default_agent_conversation")]
        conversation: AgentConversationMode,
        #[serde(default = "default_busy_policy")]
        busy_policy: BusyPolicy,
    },
    TemporaryProvider {
        provider: String,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        workspace: Option<String>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        model: Option<String>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        effort: Option<String>,
    },
}

pub type AutomationAssignments = std::collections::HashMap<String, AutomationRoleAssignment>;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AgentConversationMode {
    Current,
    FreshBackground,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum BusyPolicy {
    Wait,
    Queue,
    Skip,
    Fail,
}

fn default_agent_conversation() -> AgentConversationMode {
    AgentConversationMode::Current
}

fn default_busy_policy() -> BusyPolicy {
    BusyPolicy::Fail
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum InvocationKind {
    Manual,
    Scheduled,
    /// Started by a listener invoker reacting to an external event.
    Listener,
}

#[derive(Default, Debug, Serialize, Deserialize, Clone)]
pub struct ScheduleDefinition {
    /// "interval" | "daily" | "weekly" | "monthly" | "specific_dates" | "one_time"
    pub schedule_type: String,
    /// For interval: number of minutes between runs
    #[serde(default)]
    pub interval_minutes: Option<u32>,
    /// HH:MM in local time (used by daily, weekly, monthly, specific_dates)
    #[serde(default)]
    pub time_of_day: Option<String>,
    /// For weekly: which days (e.g. ["Mon","Tue","Fri"])
    #[serde(default)]
    pub days_of_week: Option<Vec<String>>,
    /// For weekly: repeat every N weeks (default 1)
    #[serde(default = "default_repeat_every")]
    pub repeat_every: u32,
    /// For monthly: which day(s) of the month (e.g. [1, 15])
    #[serde(default)]
    pub days_of_month: Option<Vec<u32>>,
    /// For specific_dates: list of ISO date strings ["2026-05-01", "2026-06-15"]
    #[serde(default)]
    pub specific_dates: Option<Vec<String>>,
    /// ISO8601 datetime for one_time schedules
    #[serde(default)]
    pub run_at: Option<String>,
    /// End condition: "never" | "on_date" | "after_occurrences"
    #[serde(default = "default_end_condition")]
    pub end_condition: String,
    /// ISO date (YYYY-MM-DD) for end_condition = "on_date"
    #[serde(default)]
    pub end_date: Option<String>,
    /// Count for end_condition = "after_occurrences"
    #[serde(default)]
    pub max_occurrences: Option<u32>,
    /// How many times this schedule has fired (for occurrence tracking)
    #[serde(default)]
    pub occurrence_count: u32,
    pub active: bool,
}

fn default_repeat_every() -> u32 {
    1
}
fn default_end_condition() -> String {
    "never".to_string()
}

/// A persisted automation invoker: a blueprint + invocation context (input/bindings/provider)
/// that fires on a `ScheduleDefinition` cadence.
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct AutomationSchedule {
    pub id: String,
    /// Resolves to `<home>/library/automations/<blueprint_id>.md`.
    pub blueprint_id: String,
    pub name: String,
    #[serde(default)]
    pub provider: Option<String>,
    #[serde(default)]
    pub workspace: Option<String>,
    /// Entry input params (6a `input_schema` values), passed as the run trigger.
    #[serde(default)]
    pub input: serde_json::Value,
    /// role/class -> target provider (6a bindings).
    #[serde(default)]
    pub bindings: std::collections::HashMap<String, String>,
    #[serde(default)]
    pub assignments: AutomationAssignments,
    pub schedule: ScheduleDefinition,
    #[serde(default)]
    pub next_run_epoch_ms: Option<u64>,
    #[serde(default)]
    pub paused_remaining_ms: Option<u64>,
    #[serde(default)]
    pub is_paused: bool,
    #[serde(default)]
    pub last_run_status: Option<String>,
    #[serde(default)]
    pub last_run_error: Option<String>,
    #[serde(default)]
    pub last_run_epoch_ms: Option<u64>,
}

#[cfg(test)]
mod schedule_dto_tests {
    use super::*;

    #[test]
    fn automation_schedule_round_trips_with_defaults() {
        let json = r#"{
            "id": "s1",
            "blueprint_id": "heartbeat",
            "name": "Heartbeat",
            "schedule": { "schedule_type": "interval", "interval_minutes": 60, "active": true }
        }"#;
        let s: AutomationSchedule = serde_json::from_str(json).unwrap();
        assert_eq!(s.blueprint_id, "heartbeat");
        assert!(s.provider.is_none());
        assert!(s.input.is_null() || s.input.is_object());
        assert!(s.bindings.is_empty());
        assert!(s.assignments.is_empty());
        assert!(!s.is_paused);
        let back = serde_json::to_string(&s).unwrap();
        let s2: AutomationSchedule = serde_json::from_str(&back).unwrap();
        assert_eq!(s2.id, "s1");
    }
}
